### ISSUE TITLE
oakCors() preflight NOT WORKING!!!

### ISSUE THREAD

**Issue by elnobun** [tajpouria/cors#11]
When I make a request from a `form` in the front end, I get this error: ` Response to preflight request doesn't pass access control check: It does not have HTTP ok status.
xhr.js?b50d:178 POST http://localhost:8000/login net::ERR_FAILED`.

---

I am making this request from the front end:

```
      submit() {
        axios.post("http://localhost:4000/login", this.form).catch(e => {
            this.errors = e.response.data
        })
      }
```

it works well without `this.form`

```
      submit() {
        axios.post("http://localhost:4000/login").catch(e => {
            this.errors = e.response.data
        })
      }
```

In my backend, I attached `oakCors()` in my `router.ts` as specified in the documentation:

```
router
  .options("/login", oakCors())
  .post("/login", oakCors(), AuthController.login);
```

 I also have `oakCors()` in my `main.ts` to use it

```
app.use(oakCors());
```

---

Anytime I add `this.form` as an argument, I still get the `Response to preflight` error.

**Comment by tajpouria** [comment]
@elnobun I think in order to [send form data](https://developer.mozilla.org/en-US/docs/Learn/Forms/Sending_forms_through_JavaScript) properly you should:
- Constructing form data using built-in [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData) interface
- Setting appropriate meme type header `"Content-Type": "multipart/form-data"`
- Parse multipart form-data on server using yet another module like https://github.com/deligenius/multiparser#with-oak-framework
 
Take a look at following code snippets: [Repository](https://github.com/tajpouria/node-playground/tree/master/deno-oak-demo/demos/issue-ten)

Client
```js
const addBookForm = document.getElementById("add-book-form");

addBookForm.addEventListener("submit", (event) => {
  event.preventDefault();

  const bodyFormData = new FormData(addBookForm);

  axios({
    method: "post",
    url: "http://localhost:8000/book",
    data: bodyFormData,
    headers: { "Content-Type": "multipart/form-data" },
  })
    .then(function (res) {
      console.info(res.data);

      updateBooKList();
    })
    .catch((err) => {
      console.error(err);
    });
});
```

Server
```ts
import { Application, Router } from "https://deno.land/x/oak/mod.ts";
import { oakCors } from "https://deno.land/x/cors/mod.ts";
import { multiParser } from "https://raw.githubusercontent.com/deligenius/multiparser/master/mod.ts";

interface Book {
  id: string;
  title: string;
  author: string;
}

const books = new Map<string, Book>();
books.set("1", {
  id: "1",
  title: "Frankenstein",
  author: "Mary Shelley",
});

const router = new Router();
router
  .get("/book", (context) => {
    context.response.body = Array.from(books.values());
  })
  .post("/book", async (context) => {
    const formData = await multiParser(context.request.serverRequest);

    if (
      typeof formData?.title === "string" &&
      typeof formData.author === "string"
    ) {
      const id = (books.size + 1).toString();
      const { title, author } = formData;

      books.set(id, { id, title, author });

      context.response.status = 201;
    } else {
      context.response.status = 400;
    }
  });

const app = new Application();
app.use(oakCors()); // Enable CORS for All Routes
app.use(router.routes());

console.info(`CORS-enabled web server listening on port 8000`);
await app.listen({ port: 8000 });
```

**Comment by elnobun** [comment]
> Repository

Thanks @tajpouria for your reply. Axios typically sends a POST request with custom headers to a URL. Axios automatically converts the data to JSON, so I don’t have to.

I implemented a JWT and have middleware in the backend that handles `header authorization`. It works well on the back end.


Here is my form template


```
<template>
  <div>
      <form class="flex flex-wrap justify-center p-2" @submit.prevent="submit">

        <!-- Email field-->
        <div class="py-2 w-full">
          <label for="email">
            <input
              class="p-2 w-full border rounded"
              type="email"
              placeholder="Your Email"
              v-model="form.email"
            >
          </label>
        </div>
        <!-- End email field-->

        <!-- Password Field-->
        <div class="py-2 w-full">
          <label for="password">
            <input
              class="p-2 w-full border rounded"
              type="password"
              placeholder="**********"
              v-model="form.password"
            >
          </label>
        </div>
        <!-- End password field -->

        <!-- Submit button -->
        <div class="py-2 w-full">
          <input class=" w-full px-3 py-2 bg-blue-500 shadow text-white" type="submit" value="Login">
        </div>
      </form>
  </div>
</template>

<script>
  import axios from 'axios'

  export default {
    name: 'Login',
    data() {
      return {
        form: {
          email: "",
          password: ""
        },
        errors: {},
      }
    },
    methods: {
      submit() {
        axios.post("http://localhost:4000/login", this.form).catch(e => {
            this.errors = e.response.data
        })
      }
    }
  }
</script>
```

I am not sure how to go around this problem especially for `Login forms`

**Comment by tajpouria** [comment]
@elnobun Vue app seems OK, Can you show me your AuthController.login and JWT middle-ware?

**Comment by elnobun** [comment]
@tajpouria , sure:

`AuthController`

```
import db from "../config/database.ts";
import validation from "../config/validation.ts";
import hash from "../utils/passwordHash.ts";
import token from "../utils/token.ts";

const userCollection = db.collection("users");

export default {
  async login(context: any) {
    // check validation
    const value = await validation.validateLogin(context);
    console.log(value);
    if (!value) {
      return;
    }

    // fetch user by email
    const user = await userCollection.findOne({ email: value.email });
    if (!user) {
      context.response.status = 422;
      context.response.body = { error: "Credentials do not match" };
      return;
    }

    // verify that entered password match with database
    const passwordMatch = hash.verify(user.password, value.password);
    if (!passwordMatch) {
      context.response.status = 422;
      context.response.body = { error: "Credentials do not match" };
      return;
    }
    context.response.body = token.generateToken(user._id.$oid);
  }
};
```

---

`auth.ts => middleware`

```
import token from "../utils/token.ts";

export default {
  async authorized(context: any, next: any) {
    const authorization = context.request.headers.get("authorization");
    if (!authorization) {
      context.response.status = 401;
      context.response.body = { error: "Unauthorized" };
      return;
    }

    // remove bearer from the actual token
    const headerToken = authorization.replace("Bearer ", "");

    // validate token
    const isTokenValid = await token.validateToken(headerToken);
    if (!isTokenValid) {
      context.response.status = 401;
      context.response.body = { error: "Unauthorized" };
      return;
    }
    await next();
  }
};
```

**Comment by elnobun** [comment]
I had no option than to go the `Vuejs` way, which seemed more straightforward.

1.  I have removed all instances of `oakCors()` from `main.ts` and the routes on the `server` side.

2. In the root of my `Vuejs` file, I created a `vue.config.js` file and added the `devServer` configuration to proxy API request to the backend domain. [vuejs devserver-proxy](https://cli.vuejs.org/config/#devserver-proxy)

```
module.exports = {
  devServer: {
    proxy: "http://localhost:4000"
  }
};
```

3. Then in my `form`, where I am making the `axios` API call to the server, I am doing this:

```
 submit() {
        axios.post("/login", this.form).catch(e => {
            this.errors = e.response.data
        })
      }
```

And that was all. Everything is working correctly, and I do not get the Cross-Origin Errors nor the preflight error.

Another beautiful thing is that since I put the `server URL` in the `vue.config.js` file, vuejs automatically appends the Axios API URL call to it. As such, I do not need to specify an absolute path in Axios.


Thanks for your patience @tajpouria. I think I can close this case.
